[14.0][IMP] mail_tracking: add job to delete old tracking records

henrybackman opened the pull request:
Records in `mail_tracking_email` table grow over time and increase the database size, which slows down backup and restore operations. (For example, in one customer db the `mail_tracking_email` table takes almost 2GB)

Added `autovacuum` to `mail_tracking_email` that removes by default records over 6 months old.
The deletion can be enabled with a configuration variable.
Due to possibly a large number of records to be deleted on first run, set a default limit of 5000 per run.

sebalix commented on mail_tracking/models/mail_tracking_email.py:
```diff
@@ -473,3 +473,25 @@ def event_process(self, request, post, metadata, event_type=None):
+        target_write_date = fields.Datetime.subtract(
+            fields.Datetime.now(), days=max_age_days
+        )
+        domain = [
+            ("write_date", "<", target_write_date),
+        ]
```
You should extract the domain computation in a separate method, so it's easy to override to fine-tune the domain (if we want to keep failure values for instance).

henrybackman replied:
Done

sebalix commented on mail_tracking/models/res_config_settings.py:
```diff
@@ -8,3 +8,10 @@ class ResConfigSettings(models.TransientModel):
         related="company_id.mail_tracking_show_aliases",
         readonly=False,
     )
+    enable_mail_tracking_email_deletion_job = fields.Boolean(
```
Not sure if this option is useful if we are using a dedicated cron task: we could simply archive/unarchive it. What do you think?

sebalix replied:
This option could be useful if the deletion was triggered by the `api.autovacuum` feature of Odoo.

henrybackman replied:
Switched implementation to use `api.autovacuum` and no new cron jobs required

pedrobaeza commented:
Do you have some metrics about performance degradation that leads to this action?

sebalix commented:
@pedrobaeza this action is here to reduce the DB size (to speed up backup and restore operations mainly), currently we have a DB with the `mail_tracking_value` taking almost 2GB (in the top 10 of biggest table in this DB, 8th position, 1st being `mail_message`).
About performance I checked but there isn't a real impact here (while this similar PR about notifications is reducing a lot the time needed to open the Odoo JS client: https://github.com/OCA/social/pull/1340)

pedrobaeza commented:
OK, thanks for the information. It's good to add this information initially (and include it in the commit message) for avoiding to guess what is the improvement.

It would be interesting to check if there's a way to optimize the DB schema of this table to reduce this size as well. Maybe there's a related stored or similar fields that increase the size a lot.

henrybackman commented:
Thank you @pedrobaeza for checking. I don't have anything to add to @sebalix comment but I added the relevant details to the description :+1:

henrybackman commented:
Test added to the `tests/__init__.py` and improved wording in the settings description

simahawk commented on mail_tracking/tests/test_gc_mail_tracking_email.py:
```diff
@@ -0,0 +1,80 @@
+        cls.settings = cls.env["res.config.settings"].create(
+            {"enable_old_mail_tracking_email_deletion": True}
+        )
+        cls.settings.execute()
```
Always use `set_values` instead of `execute` as it does too many things that will slow down the setup

henrybackman replied:
Changed to `set_values`

simahawk commented:
@henrybackman when you push again pls report the description in the commit msg.
Note: if you do that before opening a PR you'll get the PR description for free :wink:

sebalix commented on mail_tracking/models/mail_tracking_email.py:
```diff
@@ -473,3 +473,26 @@ def event_process(self, request, post, metadata, event_type=None):
+        return [("write_date", "<", target_write_date)]
+
+    @api.autovacuum
+    def _gc_mail_tracking_email(self, max_age_days=180, limit=5000):
```
`max_age_days` should be configurable through a system parameter as well.
We could replace the current boolean config option by an integer instead, 0 meaning nothing is deleted, or a number of retention days to enable it.

sebalix commented on mail_tracking/models/mail_tracking_email.py:
```diff
@@ -473,3 +473,26 @@ def event_process(self, request, post, metadata, event_type=None):
+            return False
+
+        domain = self._get_old_mail_tracking_email_domain(max_age_days)
+        records_to_delete = self.search(domain, limit=limit)
+        _logger.debug("Deleting %s mail.tracking.email records", len(records_to_delete))
+        return records_to_delete.unlink()
```
After testing this on a customer DB, it triggers the following error (linked to related fields):
```python
Traceback (most recent call last):
  File "/odoo/src/odoo/addons/base/models/ir_autovacuum.py", line 38, in _run_vacuum_cleaner
    func(model)
  File "/odoo/external-src/social/mail_tracking/models/mail_tracking_email.py", line 503, in _gc_mail_tracking_email
    records_to_delete.unlink()
  File "/odoo/external-src/connector/component_event/models/base.py", line 118, in unlink
    result = super(Base, self).unlink()
  File "/odoo/src/odoo/models.py", line 3501, in unlink
    self.flush()
  File "/odoo/src/odoo/models.py", line 5472, in flush
    self.recompute()
  File "/odoo/src/odoo/models.py", line 5931, in recompute
    process(field)
  File "/odoo/src/odoo/models.py", line 5918, in process
    field.recompute(existing)
  File "/odoo/src/odoo/fields.py", line 1178, in recompute
    self.compute_value(record)
  File "/odoo/src/odoo/fields.py", line 1198, in compute_value
    records._compute_field_value(self)
  File "/odoo/src/odoo/models.py", line 4079, in _compute_field_value
    odoo.fields.determine(field.compute, self)
  File "/odoo/src/odoo/fields.py", line 85, in determine
    return needle(records, *args)
  File "/odoo/src/odoo/fields.py", line 574, in _compute_related
    record[self.name] = self._process_related(value[self.related_field.name])
  File "/odoo/src/odoo/models.py", line 5702, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/odoo/src/odoo/fields.py", line 2495, in __get__
    return super().__get__(records, owner)
  File "/odoo/src/odoo/fields.py", line 1025, in __get__
    _("(Record: %s, User: %s)") % (record, env.uid),
odoo.exceptions.MissingError: Record does not exist or has been deleted.
```
I tried different things but only a raw `DELETE` query is working while being simple:
```suggestion
        records_to_delete = self.search(domain, limit=limit).exists()
        if records_to_delete:
            _logger.info("Deleting %s mail.tracking.email records", len(records_to_delete))
            self.flush()
            query = "DELETE FROM mail_tracking_email WHERE id IN %s"
            args = (tuple(records_to_delete.ids),)
            self.env.cr.execute(query, args)
            self.invalidate_cache()
```

henrybackman committed:
mail_tracking: add garbage collection

Add autovacuum to mail_tracking_email that removes old records based on new configuration variable mail_tracking_email_max_age_days.
Due to possibly a large number of records to be deleted on first run, set a default limit of 5000 per run.

sebalix commented on mail_tracking/models/mail_tracking_email.py:
```diff
@@ -473,3 +473,41 @@ def event_process(self, request, post, metadata, event_type=None):
+            max_age_days = 0
+
+        if not max_age_days > 0:
+            _logger.info("Mail tracking email deletion is disabled.")
```
As the autovacuum is run quite often, I would switch this log to DEBUG:
```suggestion
            _logger.debug("Mail tracking email deletion is disabled.")
```

pedrobaeza replied:
I don't see this as needed really. It will spam the log with no reason.

henrybackman replied:
Good points, removed it

simahawk commented:
@pedrobaeza good for you?

pedrobaeza reviewed (approved):
/ocabot merge minor
